### ISSUE TITLE
Behavior during UNREGISTER with active invocations

### ISSUE THREAD

**Issue by mbonneau** [wamp-proto/wamp-proto#283]
Behavior during `UNREGISTER` with active invocations does not appear to be defined.

Specifically: What happens when a callee sends an `UNREGISTER` message before `YIELD`ing on an `INVOCATION`?

My suggestion would be that, upon receiving an `UNREGISTER` message, the dealer must stop sending `INVOCATION` messages. If there are pending `INVOCATION`s, the dealer must wait until the callee either completes or errors the `INVOCATION` prior to sending an `UNREGISTERED` message.

**Comment by oberstet** [comment]
I agree, this is underspecified.

However, I'm not sure about your suggestion. Here is a different one:

A callee sending a YIELD for an INVOCATION on a registration _after_ sending a UNREGISTER for the respective registration is violating the protocol.

Here is how a client library would behave:

* user code calls "unregister"
* the client library will immediately cancel any currently running procedures being invoked and send ERRORs back to the dealer
* only then the client library sends the UNREGISTER

But we should discuss this and clarify - whatever we come up with.

**Comment by mbonneau** [comment]
The use case for my previous suggestion would be to allow a callee to gracefully end its service of a particular RPC without having to ERROR all pending INVOCATIONs.

Consider round-robin load balancing through multiple callees: it would be desirable for the callee to be able to request that the router stop sending INVOCATIONs while it completes pending calls so that it can shut down. This would allow dynamically increasing and decreasing capacity without causing any extra RPC errors.

The client could still have the option of sending ERRORs and then UNREGISTER.

**Comment by GregRos** [comment]
I'd like to give some input on this question because I found myself dealing with it too. 

I think @oberstet's solution is still an under-specification. A client can still receive INVOCATIONs after sending UNREGISTER, before the router has dealt with it. What should it do with those INVOCATIONs? It seems that whatever the client does, it will be in violation of protocol. It has to have no pending INVOCATIONs to send UNREGISTER, but it can't close them either. 

In addition, the client shouldn't do cleanup while it's still actively receiving jobs.

I like @mbonneau's suggestion better, but I don't see why you should delay UNREGISTERED. The client does need confirmation that the procedure has been unregistered, and it doesn't need any message to know it has finished cleanup. In addition, I don't think there is any need for a procedure to be stuck in limbo just because cleanup is being delayed.

I think a simpler solution is best:

1. Once the dealer has received UNREGISTER, the procedure becomes unregistered: future calls to the procedure will result in `no_such_procedure` and no new INVOCATIONs will be sent. The procedure can also be registered again.
2. The dealer will immediately send UNREGISTERED.
3. The callee must answer all pending invocations and the dealer will wait for it to do so.

This gives different callees flexibility to handle pending jobs in the way that best fits their use-case, while also being simpler to implement.

**Comment by LaurenceGA** [comment]
I agree with @GregRos's solution. Seems like the simplest and most sensible to me.

This issue has been open for a long time. At what point does a decision get made and the change introduced into the spec? When would it make sense to propose a pull request?

**Comment by oberstet** [comment]
I have added the "needs consensus" label removing the "needs discussion" .. I guess from above proposal, either one or a combination is sth we can agree we do wanna have. @meejah any remarks from your side?

**Comment by Jopie64** [comment]
I ran against this via #352. My 2ct:
I agree with @GregRos. We should let the callee be able to finish the pending invocations while it unregistered the procedure, so it is able to shutdown gracefully without receiving new invocations.
This would also make #352 (`REJECT`) unnecessary, cause the callee can achieve the same by (temporarily) unregistering the procedure.

**Comment by ecorm** [comment]
I'm facing this same dilemma in my C++ router implementation and will be adopting GregRos' suggestion.

**Comment by KSDaemon** [comment]
I just happened to be in this thread. I think @GregRos's solution is a good one. It would be good to describe it in spec...

**Comment by KSDaemon** [comment]
I'll remove `needs consensus` as seems most of all agree with the proposed solution.

**Comment by meejah** [comment]
Hmm, just noticed this due to the fresh activity: sounds like parallels to TCP and other protocols for "half close" (which we have to deal with in magic-wormhole too): you need a state that means "I will not send new data" before a final close. This allows both sides to "finish up" cleanly, including any "in flight" data / requests / etc.

I believe that's what has been decided here ;) just wanted to point out the parallel.
